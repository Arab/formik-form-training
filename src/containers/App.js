import React, { useState } from "react";
import { Formik } from "formik";
import { validate } from "../components/Validate";
import { blankValues, fetchedInitialValues } from "../data/Values";
import CustomFormValidate from "../components/CustomFormValidate";
import { customValidate } from "../components/CustomValidate";
import MyForm from "../components/MyForm";
import "./App.css";

function App() {
  const [initialValues, setInitialValues] = useState(blankValues);

  const fetchInitialValues = () => {
    setInitialValues({ ...fetchedInitialValues });
  };
  const handleSubmit = (values, actions) => {
    setTimeout(() => {
      alert(JSON.stringify(values, null, 2));
      actions.setSubmitting(false);
    }, 500);
  };

  const handleRender = props => {
    const form = CustomFormValidate(props, customValidate, MyForm);
    return form;
  };

  return (
    <div className="app">
      <h1>My First Formik Form </h1>
      <Formik
        fetchInitialValues={fetchInitialValues}
        initialValues={initialValues}
        validate={validate}
        onSubmit={handleSubmit}
        enableReinitialize={true}
        render={handleRender}
      />
      <button onClick={fetchInitialValues}>CLICK ME IM SERVER</button>
    </div>
  );
}
export default App;
