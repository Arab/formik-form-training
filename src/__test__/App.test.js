import React from "react";
import ReactDOM from "react-dom";
import App from "../App";
import {
  render,
  fireEvent,
  cleanup,
  waitForElement,
  wait
} from "@testing-library/react";
import "@testing-library/jest-dom/extend-expect";
import selectEvent from "react-select-event";
import userEvent from "@testing-library/user-event";

afterEach(cleanup);

it("renders without crashing", () => {
  const div = document.createElement("div");
  ReactDOM.render(<App />, div);
  ReactDOM.unmountComponentAtNode(div);
});
test("snapshot of app", () => {
  const { container } = render(<App />);
  expect(container).toMatchSnapshot();
});

test("name input exists", () => {
  const { getByPlaceholderText } = render(<App />);
  const input = getByPlaceholderText("Enter your name");
  expect(input).toBeInTheDocument();
});

test("last name input exists", () => {
  const { getByPlaceholderText } = render(<App />);
  const input = getByPlaceholderText("Enter your last name");
  expect(input).toBeInTheDocument();
});

test("age input exists", () => {
  const { getByPlaceholderText } = render(<App />);
  const input = getByPlaceholderText("Enter your age");
  expect(input).toBeInTheDocument();
});

test("male/female radio buttons exists and clicking change them", () => {
  const { getByLabelText } = render(<App />);
  const male = getByLabelText("Male");
  const female = getByLabelText("Female");
  expect(male).toBeInTheDocument();
  expect(male.checked).toBe(true);
  expect(female).toBeInTheDocument();
  expect(female.checked).toBe(false);
  fireEvent.click(female);
  expect(male.checked).toBe(false);
  expect(female.checked).toBe(true);
});

test("phone number shows corectly - every 3 digits", async () => {
  const { getByLabelText } = render(<App />);
  const actualyTypeing = (input, text) => {
    const oldValue = input.value;
    const newValue = oldValue + text;
    userEvent.type(input, newValue);
  };
  let phoneNumberInput = getByLabelText(/phone number/i);
  expect(phoneNumberInput).toBeInTheDocument();
  expect(phoneNumberInput.value).toBe("+48 ");
  userEvent.type(phoneNumberInput, "a");
  expect(phoneNumberInput.value).toBe("+48 ");
  userEvent.type(phoneNumberInput, "0.5");
  expect(phoneNumberInput.value).toBe("+48 ");
  actualyTypeing(phoneNumberInput, "000");
  expect(phoneNumberInput.value).toBe("+48 000");
  actualyTypeing(phoneNumberInput, "0");
  expect(phoneNumberInput.value).toBe("+48 000-0");
  actualyTypeing(phoneNumberInput, "00");
  expect(phoneNumberInput.value).toBe("+48 000-000");
  actualyTypeing(phoneNumberInput, "0");
  expect(phoneNumberInput.value).toBe("+48 000-000-0");
  actualyTypeing(phoneNumberInput, "00");
  expect(phoneNumberInput.value).toBe("+48 000-000-000");
  actualyTypeing(phoneNumberInput, "0");
  expect(phoneNumberInput.value).toBe("+48 000-000-000");
});

test("select destination exists", () => {
  const { getByLabelText } = render(<App />);
  const selectReact = getByLabelText(/select your destination/i);
  expect(selectReact).toBeInTheDocument();
});

test("select all button exists", () => {
  const { getByText } = render(<App />);
  const selectAllButton = getByText(/^select all$/i);
  expect(selectAllButton).toBeInTheDocument();
});

test("dietary checkboxes exists and when other is chosen there is aditional text input", async () => {
  const { getByLabelText, findByPlaceholderText } = render(<App />);
  const vegan = getByLabelText(/are you vegan?/i);
  const kosher = getByLabelText(/do you want kosher food?/i);
  const lactose = getByLabelText(/can you handle lactose?/i);
  const other = getByLabelText(/inne:/i);
  const input = findByPlaceholderText(
    "Write anything we should know about your dietary restrictions"
  );
  expect(vegan).toBeInTheDocument();
  expect(kosher).toBeInTheDocument();
  expect(lactose).toBeInTheDocument();
  expect(other).toBeInTheDocument();
  userEvent.click(other);
  await waitForElement(() => input).then(res =>
    expect(res).toBeInTheDocument()
  );
});

test("wish list label and add a thing button exists", () => {
  const { getByText } = render(<App />);
  const wishList = getByText(/things you wish to not forget:/i);
  const addAThingButton = getByText(/add a thing/i);
  expect(wishList).toBeInTheDocument();
  expect(addAThingButton).toBeInTheDocument();
});

test("Reset and Submit buttons exists", () => {
  const { getByText } = render(<App />);
  const resetButton = getByText(/reset/i);
  const submitButton = getByText("Submit");

  expect(resetButton).toBeInTheDocument();
  expect(submitButton).toBeInTheDocument();
});

//------------------- VALIDATION ---------------------------------------------------

test("name input empty validation", async () => {
  const { getByPlaceholderText, container, findByTestId } = render(<App />);
  const inputName = getByPlaceholderText("Enter your name");
  expect(container.innerHTML).not.toMatch("Required");
  fireEvent.blur(inputName);
  const validationErrors = await findByTestId("errors-firstName");
  expect(validationErrors.innerHTML).toMatch("Required");
});

test("name input to short text validation", async () => {
  const { getByPlaceholderText, container, findByTestId } = render(<App />);
  const inputName = getByPlaceholderText("Enter your name");
  expect(container.innerHTML).not.toMatch("Required");
  fireEvent.change(inputName, { target: { value: "aa" } });
  fireEvent.blur(inputName);
  const validationErrors = await findByTestId("errors-firstName");
  expect(validationErrors.innerHTML).toMatch(
    "Wysil się imie musi mieć przynajmniej 3 litery"
  );
});

test("name input is not Piotr text validation", async () => {
  const { getByPlaceholderText, container, findByTestId } = render(<App />);
  const inputName = getByPlaceholderText("Enter your name");
  expect(container.innerHTML).not.toMatch("Required");
  fireEvent.change(inputName, { target: { value: "Piotra" } });
  fireEvent.blur(inputName);
  const validationErrors = await findByTestId("errors-firstName");
  expect(validationErrors.innerHTML).toMatch("Nah kiepawe imię");
});

test("name input validation hide when proper input", async () => {
  const { getByPlaceholderText, container, findByTestId } = render(<App />);
  const inputName = getByPlaceholderText("Enter your name");
  expect(container.innerHTML).not.toMatch("Required");
  fireEvent.change(inputName, { target: { value: "Piotra" } });
  fireEvent.blur(inputName);
  const validationErrors = await findByTestId("errors-firstName");
  expect(validationErrors.innerHTML).toMatch("Nah kiepawe imię");
  fireEvent.change(inputName, { target: { value: "Piotr" } });
  await wait(() => {
    expect(container.innerHTML).not.toMatch("Nah kiepawe imię");
  });
});

test("name input validation trigered by submit button", async () => {
  const { container, findByTestId, getByText } = render(<App />);
  const submitBtn = getByText("Submit");
  expect(container.innerHTML).not.toMatch("Required");
  fireEvent.click(submitBtn);
  const validationErrors = await findByTestId("errors-firstName");
  expect(validationErrors.innerHTML).toMatch("Required");
});

test("last name input validation", async () => {
  const { container, findByTestId, getByPlaceholderText, findByText } = render(
    <App />
  );
  const input = getByPlaceholderText("Enter your last name");
  expect(container.innerHTML).not.toMatch("Last Name is Required");
  userEvent.click(input);
  fireEvent.blur(input);
  const validationError = await findByTestId("errors-lastName");
  expect(validationError).toBeInTheDocument();
  userEvent.type(input, "aa");
  const validationMessage = await findByText(
    "Nazwiska mają przynajmniej 5 znaków w jakmi świecie Ty zyjesz lel"
  );
  expect(validationMessage).toBeInTheDocument();
  userEvent.type(input, "aaaaaa");
  await wait(() => {
    expect(container.innerHTML).not.toMatch(
      "Nazwiska mają przynajmniej 5 znaków w jakmi świecie Ty zyjesz lel"
    );
    expect(container.innerHTML).not.toMatch("Last Name is Required");
  });
});

test("age input validation", async () => {
  const { getByPlaceholderText, findByTestId } = render(<App />);
  const ageInput = getByPlaceholderText("Enter your age");
  fireEvent.focus(ageInput);
  fireEvent.blur(ageInput);
  await waitForElement(() => findByTestId("errors-age")).then(res =>
    expect(res).toBeInTheDocument()
  );
});

test("phone number validation", async () => {
  const { getByLabelText, findByTestId } = render(<App />);
  const phoneNumberInput = getByLabelText(/phone number/i);
  expect(phoneNumberInput).toBeInTheDocument();
  expect(phoneNumberInput.value).toBe("+48 ");
  const typeVal = phoneNumberInput.value + "6005001";
  userEvent.type(phoneNumberInput, typeVal);
  fireEvent.blur(phoneNumberInput);
  await waitForElement(() => findByTestId("errors-phoneNumber")).then(res =>
    expect(res).toBeInTheDocument()
  );
});

test("dietary restriction validation", async () => {
  const {
    getByLabelText,
    findByTestId,
    findByPlaceholderText,
    getByTestId
  } = render(<App />);

  const vegan = getByLabelText(/are you vegan?/i);
  const kosher = getByLabelText(/do you want kosher food?/i);
  const lactose = getByLabelText(/can you handle lactose?/i);
  const other = getByLabelText(/inne:/i);

  userEvent.click(vegan);
  userEvent.click(vegan);
  fireEvent.blur(vegan);
  await waitForElement(() => findByTestId("errors-dietaryRestrictions")).then(
    res => expect(res).toBeInTheDocument()
  );
  userEvent.click(kosher);
  userEvent.click(kosher);
  fireEvent.blur(kosher);
  await waitForElement(() => findByTestId("errors-dietaryRestrictions")).then(
    res => expect(res).toBeInTheDocument()
  );
  userEvent.click(lactose);
  userEvent.click(lactose);
  fireEvent.blur(lactose);
  await waitForElement(() => findByTestId("errors-dietaryRestrictions")).then(
    res => expect(res).toBeInTheDocument()
  );
  userEvent.click(other);
  const input = findByPlaceholderText(
    "Write anything we should know about your dietary restrictions"
  );
  await waitForElement(() => input).then(res => {
    fireEvent.blur(res);
    expect(getByTestId("errors-other-value")).toBeInTheDocument();
  });
  userEvent.click(other);
  fireEvent.blur(other);
  await waitForElement(() => findByTestId("errors-dietaryRestrictions")).then(
    res => expect(res).toBeInTheDocument()
  );
});

test("todos validation", async () => {
  const { findByTestId, getByText, findByText, getAllByText } = render(<App />);
  const addAThingButton = getByText(/add a thing/i);
  fireEvent.click(addAThingButton);
  await waitForElement(() => findByText("-")).then(res => {
    fireEvent.click(res);
  });
  await waitForElement(() => findByTestId("errors-todo"))
    .then(res => {
      expect(res).toBeInTheDocument();
      expect(res.innerHTML).toBe("you need to add at least 1 thing");
    })
    .then(() => fireEvent.click(getByText(/add a thing/i)));
  await waitForElement(() => findByText("-"))
    .then(res => res.parentNode.getElementsByTagName("input")[0])
    .then(input => {
      userEvent.type(input, "cebula");
      fireEvent.blur(input);
    });
  await waitForElement(() => findByTestId("errors-todos-0")).then(res => {
    expect(res).toBeInTheDocument();
    expect(res.innerHTML).toBe(
      "No chyba nie bedziesz bral w pierwszej kolejnosci cebuli ;)"
    );
  });
  userEvent.click(getByText("+"));
  const secondPlusButton = getAllByText("+")[1];
  const secondInput = secondPlusButton.parentNode.getElementsByTagName(
    "input"
  )[0];
  userEvent.type(secondInput, "cebula");
  fireEvent.blur(secondInput);
  await waitForElement(() => findByTestId("errors-todos-1")).then(res => {
    expect(res).toBeInTheDocument();
    expect(res.innerHTML).toBe("w ogole nie zabieraj cebuli na poklad");
  });
  fireEvent.click(secondPlusButton);
  const thirdPlusButton = getAllByText("+")[2];
  const thirdInput = thirdPlusButton.parentNode.getElementsByTagName(
    "input"
  )[0];
  userEvent.type(thirdInput, "cebula");
  fireEvent.blur(thirdInput);
  await waitForElement(() => findByTestId("errors-todos-2")).then(res => {
    expect(res).toBeInTheDocument();
    expect(res.innerHTML).toBe("w ogole nie zabieraj cebuli na poklad");
  });
  fireEvent.click(thirdPlusButton);
  const fourthPlusButton = getAllByText("+")[3];
  const fourthInput = fourthPlusButton.parentNode.getElementsByTagName(
    "input"
  )[0];
  expect(fourthInput.disabled).toBe(true);
  await waitForElement(() => findByTestId("errors-todo")).then(res => {
    expect(res).toBeInTheDocument();
    expect(res.innerHTML).toBe("you can to add max 3 things");
  });
});

test("todos inputs are not empty on submit", async () => {
  const { findByTestId, getByText, getAllByText } = render(<App />);
  const addAThingButton = getByText(/add a thing/i);
  fireEvent.click(addAThingButton);
  userEvent.click(getByText("+"));
  userEvent.click(getAllByText("+")[0]);
  const submitBtn = getByText("Submit");
  fireEvent.click(submitBtn);
  await waitForElement(() => findByTestId("errors-todosEmpty")).then(res => {
    expect(res).toBeInTheDocument();
    expect(res.innerHTML).toBe("nie mozesz zostawic inputow pustych ;)");
  });
});

test("submit click fires all validation msgs", async () => {
  const { container, findByTestId, getByText } = render(<App />);
  const submitBtn = getByText("Submit");
  expect(container.innerHTML).not.toMatch("Required");
  fireEvent.click(submitBtn);
  await Promise.all([
    findByTestId("errors-firstName"),
    findByTestId("errors-lastName"),
    findByTestId("errors-age"),
    findByTestId("errors-phoneNumber"),
    findByTestId("errors-destination"),
    findByTestId("errors-dietaryRestrictions"),
    findByTestId("errors-todo")
  ]).then(errors => {
    errors.map(error => expect(error).toBeInTheDocument());
  });
});

test("validation msgs disapear after loading data from server", async () => {
  const { container, findByTestId, getByText } = render(<App />);
  const submitBtn = getByText("Submit");
  const serverBtn = getByText("CLICK ME IM SERVER");
  expect(container.innerHTML).not.toMatch("Required");
  fireEvent.click(submitBtn);
  const validationErrors = await Promise.all([
    findByTestId("errors-firstName"),
    findByTestId("errors-lastName"),
    findByTestId("errors-age"),
    findByTestId("errors-phoneNumber"),
    findByTestId("errors-destination"),
    findByTestId("errors-dietaryRestrictions"),
    findByTestId("errors-todo")
  ]).then(error => {
    error.map(err => expect(err).toBeInTheDocument());
    return error;
  });
  fireEvent.click(serverBtn);
  validationErrors.map(error => expect(error).not.toBeInTheDocument());
});

//------------------- END OF VALIDATION ---------------------------------------------------

test("reset button is disabled by default", () => {
  const { getByText } = render(<App />);
  const submitBtn = getByText("Reset");
  expect(submitBtn.disabled).toBe(true);
});

test("reset button is not disabled when form is changed", () => {
  const { getByText, getByPlaceholderText } = render(<App />);
  const submitBtn = getByText("Reset");
  const inputName = getByPlaceholderText("Enter your name");
  expect(submitBtn.disabled).toBe(true);
  // fireEvent.change(inputName, { target: { value: "aa" } });
  userEvent.type(inputName, "aa");
  expect(submitBtn.disabled).toBe(false);
});

test("test select options on react-select", async () => {
  const { getByLabelText } = render(<App />);

  const el = getByLabelText("Select your destination");
  expect(el.innerHTML).not.toMatch("Paris");
  expect(el.innerHTML).not.toMatch("London");
  await selectEvent.select(getByLabelText("Select your destination"), [
    "Paris",
    "London"
  ]);
  expect(el.innerHTML).toMatch("Paris");
  expect(el.innerHTML).toMatch("London");
});

test("test search and select option on react-select", () => {
  const { getByLabelText, getAllByText } = render(<App />);
  const el = getByLabelText("Select your destination");
  const inp = el.querySelector("input");
  userEvent.type(inp, "London");
  const sth = getAllByText("London");
  fireEvent.click(sth[1]);
  expect(getByLabelText("Select your destination").innerHTML).toMatch("London");
});

test("test search and press enter on react-select", () => {
  const { getByLabelText } = render(<App />);
  const el = getByLabelText("Select your destination");
  const inp = el.querySelector("input");
  userEvent.type(inp, "London");
  fireEvent.keyDown(el, { key: "Enter", code: 13, keyCode: 13 });
  expect(el.innerHTML).toMatch("London");
});

test("test react-select with keyboard navigation", () => {
  const { getByLabelText } = render(<App />);
  const el = getByLabelText("Select your destination");
  const inp = el.querySelector("input");
  fireEvent.focus(inp);
  fireEvent.keyDown(inp, { key: "Down", code: 40 });
  fireEvent.keyDown(inp, { key: "Down", code: 40 });
  fireEvent.keyDown(inp, { key: "Enter", code: 13, keyCode: 13 });
  expect(el.innerHTML).toMatch("London");
});

test("test select all button", () => {
  const { getByText } = render(<App />);
  const selectAllButton = getByText(/Select All/i);
  fireEvent.click(selectAllButton);
  const destinations = [
    getByText("Paris"),
    getByText("London"),
    getByText("New York"),
    getByText("Madrid"),
    getByText("Warsaw")
  ];
  destinations.map(destination => expect(destination).toBeInTheDocument());
});
