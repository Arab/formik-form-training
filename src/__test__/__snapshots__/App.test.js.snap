// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`conditional rendering other input shows up when other checkbox is checked 1`] = `
<div
  class="dietary-restrictions-wrapper"
>
  <div>
    <input
      id="isVegan"
      name="dietaryRestrictions.isVegan"
      type="checkbox"
      value=""
    />
    <label
      class="dietaryRestriction__label"
      for="isVegan"
    >
       
      Are you vegan?
    </label>
  </div>
  <div>
    <input
      id="isKosher"
      name="dietaryRestrictions.isKosher"
      type="checkbox"
      value=""
    />
    <label
      class="dietaryRestriction__label"
      for="isKosher"
    >
       
      Do you want kosher food?
    </label>
  </div>
  <div>
    <input
      id="isLactoseFree"
      name="dietaryRestrictions.isLactoseFree"
      type="checkbox"
      value=""
    />
    <label
      class="dietaryRestriction__label"
      for="isLactoseFree"
    >
       
      Can you handle Lactose?
    </label>
  </div>
  <div>
    <input
      id="other"
      name="dietaryRestrictions.other.isOther"
      type="checkbox"
      value=""
    />
    <label
      class="dietaryRestriction__label"
      for="other"
    >
       
      inne:
    </label>
  </div>
  <div />
</div>
`;

exports[`conditional rendering other input shows up when other checkbox is checked 2`] = `
<div
  class="dietary-restrictions-wrapper"
>
  <div>
    <input
      id="isVegan"
      name="dietaryRestrictions.isVegan"
      type="checkbox"
      value=""
    />
    <label
      class="dietaryRestriction__label"
      for="isVegan"
    >
       
      Are you vegan?
    </label>
  </div>
  <div>
    <input
      id="isKosher"
      name="dietaryRestrictions.isKosher"
      type="checkbox"
      value=""
    />
    <label
      class="dietaryRestriction__label"
      for="isKosher"
    >
       
      Do you want kosher food?
    </label>
  </div>
  <div>
    <input
      id="isLactoseFree"
      name="dietaryRestrictions.isLactoseFree"
      type="checkbox"
      value=""
    />
    <label
      class="dietaryRestriction__label"
      for="isLactoseFree"
    >
       
      Can you handle Lactose?
    </label>
  </div>
  <div>
    <input
      id="other"
      name="dietaryRestrictions.other.isOther"
      type="checkbox"
      value=""
    />
    <label
      class="dietaryRestriction__label"
      for="other"
    >
       
      inne:
    </label>
  </div>
  <div>
    <div
      class="other__input"
    >
      <input
        class="other"
        name="dietaryRestrictions.other.value"
        placeholder="Write anything we should know about your dietary restrictions"
        type="text"
        value=""
      />
    </div>
  </div>
</div>
`;

exports[`snapshot of app 1`] = `
<div>
  <div
    class="app"
  >
    <h1>
      My First Formik Form 
    </h1>
    <form>
      <input
        class="input"
        name="firstName"
        placeholder="Enter your name"
        type="text"
        value=""
      />
      <input
        class="input"
        name="lastName"
        placeholder="Enter your last name"
        type="text"
        value=""
      />
      <input
        class="input"
        name="age"
        placeholder="Enter your age"
        type="number"
        value="0"
      />
      <div
        class="gender-wrapper"
      >
        <input
          checked=""
          id="male"
          name="gender"
          type="radio"
          value="male"
        />
        <label
          for="male"
        >
           Male
        </label>
        <input
          id="female"
          name="gender"
          type="radio"
          value="female"
        />
        <label
          for="female"
        >
           Female
        </label>
      </div>
      <label>
        Phone Number:
         
        <input
          class="phoneNumber"
          name="phoneNumber"
          type="text"
          value="+48 "
        />
      </label>
      <div
        style="margin: 1rem 0px;"
      >
        <label
          for="color"
        >
          Select your destination 
        </label>
        <div
          class=" css-2b097c-container"
          id="color"
        >
          <div
            class=" css-yk16xz-control"
          >
            <div
              class=" css-1hwfws3"
            >
              <div
                class=" css-1wa3eu0-placeholder"
              >
                Select...
              </div>
              <div
                class="css-1g6gooi"
              >
                <div
                  class=""
                  style="display: inline-block;"
                >
                  <input
                    aria-autocomplete="list"
                    autocapitalize="none"
                    autocomplete="off"
                    autocorrect="off"
                    id="react-select-2-input"
                    spellcheck="false"
                    style="box-sizing: content-box; width: 2px; border: 0px; font-size: inherit; opacity: 1; outline: 0; padding: 0px;"
                    tabindex="0"
                    type="text"
                    value=""
                  />
                  <div
                    style="position: absolute; top: 0px; left: 0px; visibility: hidden; height: 0px; overflow: scroll; white-space: pre; font-size: inherit; font-family: -webkit-small-control; letter-spacing: normal; text-transform: none;"
                  />
                </div>
              </div>
            </div>
            <div
              class=" css-1wy0on6"
            >
              <span
                class=" css-1okebmr-indicatorSeparator"
              />
              <div
                aria-hidden="true"
                class=" css-tlfecz-indicatorContainer"
              >
                <svg
                  aria-hidden="true"
                  class="css-19bqh2r"
                  focusable="false"
                  height="20"
                  viewBox="0 0 20 20"
                  width="20"
                >
                  <path
                    d="M4.516 7.548c0.436-0.446 1.043-0.481 1.576 0l3.908 3.747 3.908-3.747c0.533-0.481 1.141-0.446 1.574 0 0.436 0.445 0.408 1.197 0 1.615-0.406 0.418-4.695 4.502-4.695 4.502-0.217 0.223-0.502 0.335-0.787 0.335s-0.57-0.112-0.789-0.335c0 0-4.287-4.084-4.695-4.502s-0.436-1.17 0-1.615z"
                  />
                </svg>
              </div>
            </div>
          </div>
        </div>
      </div>
      <div
        style="margin: 1rem 0px;"
      >
        <button
          type="button"
        >
          Select All
        </button>
      </div>
      <div
        class="dietary-restrictions-wrapper"
      >
        <div>
          <input
            id="isVegan"
            name="dietaryRestrictions.isVegan"
            type="checkbox"
            value=""
          />
          <label
            class="dietaryRestriction__label"
            for="isVegan"
          >
             
            Are you vegan?
          </label>
        </div>
        <div>
          <input
            id="isKosher"
            name="dietaryRestrictions.isKosher"
            type="checkbox"
            value=""
          />
          <label
            class="dietaryRestriction__label"
            for="isKosher"
          >
             
            Do you want kosher food?
          </label>
        </div>
        <div>
          <input
            id="isLactoseFree"
            name="dietaryRestrictions.isLactoseFree"
            type="checkbox"
            value=""
          />
          <label
            class="dietaryRestriction__label"
            for="isLactoseFree"
          >
             
            Can you handle Lactose?
          </label>
        </div>
        <div>
          <input
            id="other"
            name="dietaryRestrictions.other.isOther"
            type="checkbox"
            value=""
          />
          <label
            class="dietaryRestriction__label"
            for="other"
          >
             
            inne:
          </label>
        </div>
        <div />
      </div>
      <div>
        <div
          class="todosLabel"
        >
          Things you wish to not forget:
        </div>
        <div>
          <button
            type="button"
          >
            Add a thing
          </button>
        </div>
      </div>
      <div>
        <button
          class="outline"
          disabled=""
          type="button"
        >
          Reset
        </button>
        <button
          type="submit"
        >
          Submit
        </button>
      </div>
      <div
        style="margin: 1rem 0px;"
      >
        <pre
          style="background: rgb(246, 248, 250); font-size: .65rem; padding: .5rem;"
        >
          <strong>
            props
          </strong>
           = 
          {
  "values": {
    "firstName": "",
    "lastName": "",
    "age": 0,
    "gender": "male",
    "phoneNumber": "+48 ",
    "destination": [],
    "dietaryRestrictions": {
      "isVegan": false,
      "isKosher": false,
      "isLactoseFree": false,
      "other": {
        "isOther": false,
        "value": ""
      }
    },
    "todos": [],
    "todosEmpty": ""
  },
  "errors": {},
  "touched": {},
  "isSubmitting": false,
  "isValidating": false,
  "submitCount": 0,
  "dirty": false,
  "isValid": false,
  "initialValues": {
    "firstName": "",
    "lastName": "",
    "age": 0,
    "gender": "male",
    "phoneNumber": "+48 ",
    "destination": [],
    "dietaryRestrictions": {
      "isVegan": false,
      "isKosher": false,
      "isLactoseFree": false,
      "other": {
        "isOther": false,
        "value": ""
      }
    },
    "todos": [],
    "todosEmpty": ""
  },
  "validateOnChange": true,
  "validateOnBlur": true
}
        </pre>
      </div>
    </form>
    <button>
      CLICK ME IM SERVER
    </button>
  </div>
</div>
`;
